src/track/beatutils.cpp

```diff
@@ -12,37 +12,314 @@
 #include <QMap>
 
 #include "track/beatutils.h"
+#include "track/beatstats.h"
 #include "util/math.h"
 
+namespace {
+
 // we are generous and assume the global_BPM to be at most 0.05 BPM far away
 // from the correct one
-#define BPM_ERROR 0.05
+constexpr double kMaxBpmError = 0.05;
+constexpr double kMaxSecsPhaseError = 0.025; //25ms
 
 // the raw beatgrid is divided into blocks of size N from which the local bpm is
 // computed. Tweaked from 8 to 12 which improves the BPM accuracy for 'problem songs'.
-#define N 12
+constexpr int kBeatsToCountTempo = 12;
 
 static bool sDebug = false;
 
 const double kCorrectBeatLocalBpmEpsilon = 0.05; //0.2;
 const int kHistogramDecimalPlaces = 2;
 const double kHistogramDecimalScale = pow(10.0, kHistogramDecimalPlaces);
 const double kBpmFilterTolerance = 1.0;
+constexpr double kMaxDiffSameBpm = 0.25;
+
+bool checkIfErrorGrowing(double previous, double current) {
+
+    if (current < 0 && previous > 0) {
+        return false;
+    } else if (current > 0 && previous < 0) {
+        return true;
+    } else {
+        return fabs(current) > fabs(previous)
+                ? true : false;
+    }
+}
+
+} //namespece
+
+QMap<int, double> BeatUtils::findTempoChanges(
+        QMap<double, int> tempoFrequency, QList<double> tempoList) {
+
+    auto sortedTempoList = tempoList;
+    std::sort(sortedTempoList.begin(), sortedTempoList.end());
+    // We have to make sure we have odd numbers
+    if (!(sortedTempoList.size() % 2) and sortedTempoList.size() > 1) {
+        sortedTempoList.pop_back();
+    }
+    // Since we use the median as a guess first and last tempo
+    // And it can not have values outside from tempoFrequency
+    const double medianBpm = computeSampleMedian(sortedTempoList);
+    // Forming a meter perception takes a few seconds, so we assume sections of consistent 
+    // metrical structure to be at least around 10s long. So we use a window of the double
```

Comment:
Do not get this too, it's just the arbitrary reasoning for choosing the window instead of any other? Or do you mean specifically about the double of that?

Reply:
I think we just need a longer comments. 
In all the comments be sure not to explain the code we can read anyway, but describe why it is there. 
In this case the advantages against the obvious solution, that was abandoned are interesting.